code/modules/power/turbine/turbine.dm

```diff
@@ -0,0 +1,557 @@
+			if(!do_after(user, 2 SECONDS, src))
+				return
+			user.transferItemToLoc(rotor_part, src)
+			user.put_in_hands(installed_part)
+			installed_part = rotor_part
+			rotor_part_efficiency = rotor_part.part_efficiency
+			calculate_parts_limits()
+			balloon_alert(user, "replaced part with a better one")
+			return
+
+		balloon_alert(user, "already installed")
+
+	return ..()
+
+/obj/machinery/power/turbine/core_rotor/on_deconstruction()
+	if(all_parts_connected)
+		deactivate_parts()
+	return ..()
+
+/obj/machinery/power/turbine/core_rotor/proc/calculate_parts_limits()
+	compressor_part_efficiency = compressor.installed_part.part_efficiency
+	stator_part_efficiency = turbine.installed_part.part_efficiency
+	rotor_part_efficiency = installed_part.part_efficiency
+
+	max_allowed_rpm = (compressor.installed_part.max_rpm + turbine.installed_part.max_rpm + installed_part.max_rpm) / 3
+	max_allowed_temperature = (compressor.installed_part.max_temperature + turbine.installed_part.max_temperature + installed_part.max_temperature) / 3
+
+/obj/machinery/power/turbine/core_rotor/proc/calculate_damage_done(temperature)
+	damage_archived = damage
+	var/temperature_difference = temperature - max_allowed_temperature
+	var/damage_done = round(log(90, max(temperature_difference, 1)), 0.5)
+
+	damage = max(damage + damage_done * 0.5, 0)
+	damage = min(damage_archived + TURBINE_MAX_TAKEN_DAMAGE, damage)
+	if(temperature_difference < 0)
+		damage = max(damage - TURBINE_DAMAGE_HEALING, 0)
+
+	if((damage - damage_archived >= 2 || damage > TURBINE_DAMAGE_ALARM_START) && COOLDOWN_FINISHED(src, turbine_damage_alert))
+		damage_alert(damage_done)
+
+/obj/machinery/power/turbine/core_rotor/proc/damage_alert(damage_done)
+	COOLDOWN_START(src, turbine_damage_alert, max(round(TURBINE_DAMAGE_ALARM_START - damage_done), 5) SECONDS)
+
+	var/integrity = damage / 500
+	integrity = max(round(100 - integrity * 100, 0.01), 0)
+
+	if(integrity <= 0)
+		failure()
+		return
+
+	radio.talk_into(src, "Warning, turbine at [get_area_name(src)] taking damage, current integrity at [integrity]%!", engineering_channel)
+
+/obj/machinery/power/turbine/core_rotor/proc/failure()
+	deactivate_parts()
+	if(rpm < 35000)
+		explosion(src, 0, 1, 4)
+		return
+	if(rpm < 87500)
+		explosion(src, 0, 2, 6)
+		return
+	if(rpm < 220000)
+		explosion(src, 1, 3, 7)
+		return
+	if(rpm < 550000)
+		explosion(src, 2, 5, 7)
+
+/obj/machinery/power/turbine/core_rotor/process_atmos()
+
+	if(!active || !all_parts_connected)
+		return
+
+	var/datum/gas_mixture/input_turf_mixture = input_turf.air
+
+	if(!input_turf_mixture || !input_turf_mixture.gases)
+		return
+
+	calculate_damage_done(input_turf_mixture.temperature)
+
+	var/compressor_work = input_turf_mixture.total_moles() * R_IDEAL_GAS_EQUATION * input_turf_mixture.temperature * log(input_turf_mixture.volume / compressor_mixture.volume) * TURBINE_WORK_CONVERSION_MULTIPLIER
+	input_turf.air.pump_gas_to(compressor_mixture, input_turf.air.return_pressure())
```

Comment:
I don't think you ever actually activate the input turf, you should do that down at the bottom with the output turf

Reply:
Ye I forgot, will do it

Reply:
never mind, i already did it just in the line under that one